[2025-03-04T20:29:03.845+0000] {local_task_job_runner.py:123} INFO - ::group::Pre task execution logs
[2025-03-04T20:29:03.922+0000] {taskinstance.py:2614} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: fraud_detection_training.execute_training manual__2025-03-04T20:28:58.654145+00:00 [queued]>
[2025-03-04T20:29:03.936+0000] {taskinstance.py:2614} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: fraud_detection_training.execute_training manual__2025-03-04T20:28:58.654145+00:00 [queued]>
[2025-03-04T20:29:03.936+0000] {taskinstance.py:2867} INFO - Starting attempt 1 of 1
[2025-03-04T20:29:03.965+0000] {taskinstance.py:2890} INFO - Executing <Task(PythonOperator): execute_training> on 2025-03-04 20:28:58.654145+00:00
[2025-03-04T20:29:03.971+0000] {warnings.py:112} WARNING - /home/***/.local/lib/python3.12/site-packages/***/task/task_runner/standard_task_runner.py:70: DeprecationWarning: This process (pid=9561) is multi-threaded, use of fork() may lead to deadlocks in the child.
  pid = os.fork()

[2025-03-04T20:29:03.974+0000] {standard_task_runner.py:72} INFO - Started process 9562 to run task
[2025-03-04T20:29:03.975+0000] {standard_task_runner.py:104} INFO - Running: ['***', 'tasks', 'run', 'fraud_detection_training', 'execute_training', 'manual__2025-03-04T20:28:58.654145+00:00', '--job-id', '31', '--raw', '--subdir', 'DAGS_FOLDER/fraud_detection_training_dag.py', '--cfg-path', '/tmp/tmpskjd5k6w']
[2025-03-04T20:29:03.976+0000] {standard_task_runner.py:105} INFO - Job 31: Subtask execute_training
[2025-03-04T20:29:04.173+0000] {task_command.py:467} INFO - Running <TaskInstance: fraud_detection_training.execute_training manual__2025-03-04T20:28:58.654145+00:00 [running]> on host b7f28f610775
[2025-03-04T20:29:04.570+0000] {taskinstance.py:3134} INFO - Exporting env vars: AIRFLOW_CTX_DAG_OWNER='datamasterylab.com' AIRFLOW_CTX_DAG_ID='fraud_detection_training' AIRFLOW_CTX_TASK_ID='execute_training' AIRFLOW_CTX_EXECUTION_DATE='2025-03-04T20:28:58.654145+00:00' AIRFLOW_CTX_TRY_NUMBER='1' AIRFLOW_CTX_DAG_RUN_ID='manual__2025-03-04T20:28:58.654145+00:00'
[2025-03-04T20:29:04.571+0000] {taskinstance.py:732} INFO - ::endgroup::
[2025-03-04T20:29:07.982+0000] {warnings.py:112} WARNING - /opt/***/dags/fraud_detection_training.py:15: UserWarning: The numpy.array_api submodule is still experimental. See NEP 47.
  from numpy.array_api import astype

[2025-03-04T20:29:08.061+0000] {fraud_detection_training_dag.py:23} INFO - Initializing fraud detection training
[2025-03-04T20:29:08.066+0000] {fraud_detection_training.py:59} INFO - Configuration loaded successfully
[2025-03-04T20:29:08.232+0000] {fraud_detection_training.py:84} INFO - Minio connection verified. Buckets: ['mlflow']
[2025-03-04T20:29:08.315+0000] {fraud_detection_training.py:171} INFO - Starting model training process
[2025-03-04T20:29:08.315+0000] {fraud_detection_training.py:97} INFO - Connecting to kafka topic transactions
[2025-03-04T20:29:08.375+0000] {conn.py:384} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=bootstrap-0 host=pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connecting> [IPv4 ('18.169.210.18', 9092)]>: connecting to pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 [('18.169.210.18', 9092) IPv4]
[2025-03-04T20:29:08.376+0000] {conn.py:1219} INFO - Probing node bootstrap-0 broker version
[2025-03-04T20:29:08.467+0000] {conn.py:480} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=bootstrap-0 host=pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <handshake> [IPv4 ('18.169.210.18', 9092)]>: Loading system default SSL CAs from ('/usr/lib/ssl/cert.pem', '/usr/lib/ssl/certs', 'SSL_CERT_FILE', '/usr/lib/ssl/cert.pem', 'SSL_CERT_DIR', '/usr/lib/ssl/certs')
[2025-03-04T20:29:08.757+0000] {conn.py:665} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=bootstrap-0 host=pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('18.169.210.18', 9092)]>: Authenticated as XGISOSVJLGK43N6I via PLAIN
[2025-03-04T20:29:08.759+0000] {conn.py:449} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=bootstrap-0 host=pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('18.169.210.18', 9092)]>: Connection complete.
[2025-03-04T20:29:08.906+0000] {conn.py:1282} INFO - Broker version identified as 2.6.0
[2025-03-04T20:29:08.907+0000] {conn.py:1283} INFO - Set configuration api_version=(2, 6, 0) to skip auto check_version requests on startup
[2025-03-04T20:29:08.910+0000] {consumer.py:118} WARNING - group_id is None: disabling auto-commit.
[2025-03-04T20:29:08.911+0000] {subscription_state.py:171} INFO - Updating subscribed topics to: ('transactions',)
[2025-03-04T20:29:09.002+0000] {subscription_state.py:257} INFO - Updated partition assignment: [('transactions', 0), ('transactions', 1), ('transactions', 2), ('transactions', 3), ('transactions', 4), ('transactions', 5)]
[2025-03-04T20:29:09.025+0000] {conn.py:384} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=4 host=b4-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connecting> [IPv4 ('18.135.141.220', 9092)]>: connecting to b4-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 [('18.135.141.220', 9092) IPv4]
[2025-03-04T20:29:09.129+0000] {conn.py:480} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=4 host=b4-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <handshake> [IPv4 ('18.135.141.220', 9092)]>: Loading system default SSL CAs from ('/usr/lib/ssl/cert.pem', '/usr/lib/ssl/certs', 'SSL_CERT_FILE', '/usr/lib/ssl/cert.pem', 'SSL_CERT_DIR', '/usr/lib/ssl/certs')
[2025-03-04T20:29:09.605+0000] {conn.py:665} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=4 host=b4-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('18.135.141.220', 9092)]>: Authenticated as XGISOSVJLGK43N6I via PLAIN
[2025-03-04T20:29:09.607+0000] {conn.py:449} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=4 host=b4-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('18.135.141.220', 9092)]>: Connection complete.
[2025-03-04T20:29:09.607+0000] {conn.py:928} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=bootstrap-0 host=pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('18.169.210.18', 9092)]>: Closing connection. 
[2025-03-04T20:29:09.736+0000] {conn.py:384} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=1 host=b1-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connecting> [IPv4 ('13.41.38.144', 9092)]>: connecting to b1-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 [('13.41.38.144', 9092) IPv4]
[2025-03-04T20:29:09.842+0000] {conn.py:480} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=1 host=b1-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <handshake> [IPv4 ('13.41.38.144', 9092)]>: Loading system default SSL CAs from ('/usr/lib/ssl/cert.pem', '/usr/lib/ssl/certs', 'SSL_CERT_FILE', '/usr/lib/ssl/cert.pem', 'SSL_CERT_DIR', '/usr/lib/ssl/certs')
[2025-03-04T20:29:10.390+0000] {conn.py:665} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=1 host=b1-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('13.41.38.144', 9092)]>: Authenticated as XGISOSVJLGK43N6I via PLAIN
[2025-03-04T20:29:10.393+0000] {conn.py:449} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=1 host=b1-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('13.41.38.144', 9092)]>: Connection complete.
[2025-03-04T20:29:10.585+0000] {conn.py:384} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=2 host=b2-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connecting> [IPv4 ('13.43.33.231', 9092)]>: connecting to b2-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 [('13.43.33.231', 9092) IPv4]
[2025-03-04T20:29:10.694+0000] {conn.py:480} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=2 host=b2-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <handshake> [IPv4 ('13.43.33.231', 9092)]>: Loading system default SSL CAs from ('/usr/lib/ssl/cert.pem', '/usr/lib/ssl/certs', 'SSL_CERT_FILE', '/usr/lib/ssl/cert.pem', 'SSL_CERT_DIR', '/usr/lib/ssl/certs')
[2025-03-04T20:29:11.092+0000] {conn.py:665} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=2 host=b2-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('13.43.33.231', 9092)]>: Authenticated as XGISOSVJLGK43N6I via PLAIN
[2025-03-04T20:29:11.094+0000] {conn.py:449} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=2 host=b2-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('13.43.33.231', 9092)]>: Connection complete.
[2025-03-04T20:29:11.230+0000] {conn.py:384} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=5 host=b5-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connecting> [IPv4 ('35.177.218.127', 9092)]>: connecting to b5-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 [('35.177.218.127', 9092) IPv4]
[2025-03-04T20:29:11.340+0000] {conn.py:480} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=5 host=b5-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <handshake> [IPv4 ('35.177.218.127', 9092)]>: Loading system default SSL CAs from ('/usr/lib/ssl/cert.pem', '/usr/lib/ssl/certs', 'SSL_CERT_FILE', '/usr/lib/ssl/cert.pem', 'SSL_CERT_DIR', '/usr/lib/ssl/certs')
[2025-03-04T20:29:11.819+0000] {conn.py:665} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=5 host=b5-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('35.177.218.127', 9092)]>: Authenticated as XGISOSVJLGK43N6I via PLAIN
[2025-03-04T20:29:11.821+0000] {conn.py:449} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=5 host=b5-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('35.177.218.127', 9092)]>: Connection complete.
[2025-03-04T20:29:13.212+0000] {conn.py:384} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=3 host=b3-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connecting> [IPv4 ('13.43.118.221', 9092)]>: connecting to b3-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 [('13.43.118.221', 9092) IPv4]
[2025-03-04T20:29:13.321+0000] {conn.py:480} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=3 host=b3-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <handshake> [IPv4 ('13.43.118.221', 9092)]>: Loading system default SSL CAs from ('/usr/lib/ssl/cert.pem', '/usr/lib/ssl/certs', 'SSL_CERT_FILE', '/usr/lib/ssl/cert.pem', 'SSL_CERT_DIR', '/usr/lib/ssl/certs')
[2025-03-04T20:29:13.810+0000] {conn.py:665} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=3 host=b3-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('13.43.118.221', 9092)]>: Authenticated as XGISOSVJLGK43N6I via PLAIN
[2025-03-04T20:29:13.811+0000] {conn.py:449} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=3 host=b3-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('13.43.118.221', 9092)]>: Connection complete.
[2025-03-04T20:29:13.963+0000] {conn.py:384} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=0 host=b0-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connecting> [IPv4 ('35.179.249.93', 9092)]>: connecting to b0-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 [('35.179.249.93', 9092) IPv4]
[2025-03-04T20:29:14.068+0000] {conn.py:480} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=0 host=b0-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <handshake> [IPv4 ('35.179.249.93', 9092)]>: Loading system default SSL CAs from ('/usr/lib/ssl/cert.pem', '/usr/lib/ssl/certs', 'SSL_CERT_FILE', '/usr/lib/ssl/cert.pem', 'SSL_CERT_DIR', '/usr/lib/ssl/certs')
[2025-03-04T20:29:14.498+0000] {conn.py:665} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=0 host=b0-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('35.179.249.93', 9092)]>: Authenticated as XGISOSVJLGK43N6I via PLAIN
[2025-03-04T20:29:14.501+0000] {conn.py:449} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=0 host=b0-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <authenticating> [IPv4 ('35.179.249.93', 9092)]>: Connection complete.
[2025-03-04T20:29:31.694+0000] {conn.py:928} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=4 host=b4-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('18.135.141.220', 9092)]>: Closing connection. 
[2025-03-04T20:29:31.708+0000] {conn.py:928} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=1 host=b1-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('13.41.38.144', 9092)]>: Closing connection. 
[2025-03-04T20:29:31.709+0000] {fetcher.py:782} INFO - Fetch to node 1 failed: Cancelled: <BrokerConnection client_id=kafka-python-2.0.5, node_id=1 host=b1-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('13.41.38.144', 9092)]>
[2025-03-04T20:29:31.710+0000] {conn.py:928} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=2 host=b2-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('13.43.33.231', 9092)]>: Closing connection. 
[2025-03-04T20:29:31.710+0000] {conn.py:928} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=5 host=b5-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('35.177.218.127', 9092)]>: Closing connection. 
[2025-03-04T20:29:31.712+0000] {fetcher.py:782} INFO - Fetch to node 5 failed: Cancelled: <BrokerConnection client_id=kafka-python-2.0.5, node_id=5 host=b5-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('35.177.218.127', 9092)]>
[2025-03-04T20:29:31.712+0000] {conn.py:928} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=3 host=b3-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('13.43.118.221', 9092)]>: Closing connection. 
[2025-03-04T20:29:31.713+0000] {fetcher.py:782} INFO - Fetch to node 3 failed: Cancelled: <BrokerConnection client_id=kafka-python-2.0.5, node_id=3 host=b3-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('13.43.118.221', 9092)]>
[2025-03-04T20:29:31.713+0000] {conn.py:928} INFO - <BrokerConnection client_id=kafka-python-2.0.5, node_id=0 host=b0-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('35.179.249.93', 9092)]>: Closing connection. 
[2025-03-04T20:29:31.713+0000] {fetcher.py:782} INFO - Fetch to node 0 failed: Cancelled: <BrokerConnection client_id=kafka-python-2.0.5, node_id=0 host=b0-pkc-41wq6.eu-west-2.aws.confluent.cloud:9092 <connected> [IPv4 ('35.179.249.93', 9092)]>
[2025-03-04T20:29:33.234+0000] {fraud_detection_training.py:125} INFO - Kafka data read successfully with fraud rate: 0.38%
[2025-03-04T20:29:40.544+0000] {fraud_detection_training.py:250} INFO - Starting hyperparameter tuning...
[2025-03-04T20:29:40.549+0000] {warnings.py:112} WARNING - /home/***/.local/lib/python3.12/site-packages/joblib/parallel.py:1359: UserWarning: Loky-backed parallel loops cannot be called in a multiprocessing, setting n_jobs=1
  n_jobs = self._backend.configure(n_jobs=self.n_jobs, parallel=self,

[2025-03-04T20:32:59.584+0000] {fraud_detection_training.py:254} INFO - Best hyperparameters: {'classifier__subsample': 0.6, 'classifier__reg_alpha': 0, 'classifier__max_depth': 5, 'classifier__learning_rate': 0.1, 'classifier__gamma': 0.3, 'classifier__colsample_bytree': 0.8}
[2025-03-04T20:33:00.041+0000] {fraud_detection_training.py:260} INFO - Optimal threshold determined: 0.8057
[2025-03-04T20:33:01.499+0000] {credentials.py:1147} INFO - Found credentials in environment variables.
[2025-03-04T20:33:02.635+0000] {warnings.py:112} WARNING - /home/***/.local/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(

[2025-03-04T20:33:08.680+0000] {logging_mixin.py:190} WARNING - Registered model 'fraud_detection_model' already exists. Creating a new version of this model...
[2025-03-04T20:33:08.720+0000] {logging_mixin.py:190} WARNING - 2025/03/04 20:33:08 INFO mlflow.store.model_registry.abstract_store: Waiting up to 300 seconds for model version to finish creation. Model name: fraud_detection_model, version 4
[2025-03-04T20:33:08.721+0000] {logging_mixin.py:190} WARNING - Created version '4' of model 'fraud_detection_model'.
[2025-03-04T20:33:08.778+0000] {fraud_detection_training.py:322} INFO - Training successfully completed with metrics: {'auc_pr': 0.4392693641987746, 'precision': 0.9655172413793104, 'recall': 0.42748091603053434, 'f1': 0.5925925925925926, 'threshold': 0.8057419061660767}
[2025-03-04T20:33:08.791+0000] {logging_mixin.py:190} INFO - 🏃 View run persistent-zebra-8 at: http://mlflow-server:5500/#/experiments/1/runs/1888a3af8b63447fb9247b9bf0ae9c84
[2025-03-04T20:33:08.792+0000] {logging_mixin.py:190} INFO - 🧪 View experiment at: http://mlflow-server:5500/#/experiments/1
[2025-03-04T20:33:08.883+0000] {python.py:240} INFO - Done. Returned value was: {'status': 'success', 'precision': {'auc_pr': 0.4392693641987746, 'precision': 0.9655172413793104, 'recall': 0.42748091603053434, 'f1': 0.5925925925925926, 'threshold': 0.8057419061660767}}
[2025-03-04T20:33:08.997+0000] {taskinstance.py:341} INFO - ::group::Post task execution logs
[2025-03-04T20:33:08.998+0000] {taskinstance.py:353} INFO - Marking task as SUCCESS. dag_id=fraud_detection_training, task_id=execute_training, run_id=manual__2025-03-04T20:28:58.654145+00:00, execution_date=20250304T202858, start_date=20250304T202903, end_date=20250304T203308
[2025-03-04T20:33:09.164+0000] {local_task_job_runner.py:266} INFO - Task exited with return code 0
[2025-03-04T20:33:09.258+0000] {taskinstance.py:3901} INFO - 1 downstream tasks scheduled from follow-on schedule check
[2025-03-04T20:33:09.262+0000] {local_task_job_runner.py:245} INFO - ::endgroup::
